astropy/time/formats.py

```diff
@@ -2175,6 +2176,204 @@ def value(self):
+    1.666666667d
+    """
+
+    name = "quantity_str"
+
+    subfmts = (
+        ("multi", None, None),
+        ("yr", None, None),
+        ("d", None, None),
+        ("hr", None, None),
+        ("min", None, None),
+        ("s", None, None),
+    )
+
+    # Regex to parse "1.02yr 2.2d 3.12hr 4.322min 5.6s" where each element is optional
+    # but the order is fixed. Each element is a float with optional exponent. Each
+    # element is named.
+    re_float = r"(\d+(\.\d*)?|\.\d+)([eE][-+]?\d+)?"
+    re_ydhms = re.compile(
+        rf"""^ \s*
+        (?P<sign>[-+])? \s*  # Optional sign
+        (?=[^-+\s])  # At least one character which is not a sign or whitespace
+        ((?P<yr>{re_float}) \s* yr \s*)?
+        ((?P<d>{re_float}) \s* d \s*)?
+        ((?P<hr>{re_float}) \s* hr \s*)?
+        ((?P<min>{re_float}) \s* min \s*)?
+        ((?P<s>{re_float}) \s* s)?
+        \s* $
+        """,
+        re.VERBOSE,
+    )
+
+    def _check_val_type(self, val1, val2):
+        if val1.dtype.kind not in ("S", "U") and val1.size:
+            raise TypeError(f"Input values for {self.name} class must be strings")
+        if val2 is not None:
+            raise ValueError(
+                f"{self.name} objects do not accept a val2 but you provided {val2}"
+            )
+        return val1, None
+
+    def parse_string(self, timestr):
+        """Read time from a single string"""
+        # Datetime components required for conversion to JD by ERFA, along
+        # with the default values.
+        components = ("yr", "d", "hr", "min", "s")
+
+        if (match := self.re_ydhms.match(timestr)) is None:
+            raise ValueError(
+                f"Time delta '{timestr}' does not match {self.name} format"
+            )
+
+        tm = match.groupdict()
+        vals = [float(tm[component] or 0.0) for component in components]
+        if tm["sign"] == "-":
+            vals = [-val for val in vals]
+
+        return vals
+
+    def set_jds(self, val1, val2):
+        """Parse the time strings contained in val1 and get jd1, jd2."""
+        # Be liberal in what we accept: convert bytes to ascii.
+        # Here .item() is needed for arrays with entries of unequal length,
+        # to strip trailing 0 bytes.
+        to_string = (
+            str if val1.dtype.kind == "U" else lambda x: str(x.item(), encoding="ascii")
+        )
+        iterator = np.nditer(
+            [val1, None, None, None, None, None],
+            flags=["zerosize_ok"],
+            op_dtypes=[None] + 5 * [np.double],
+        )
+        for val, yr, day, hr, min, sec in iterator:
+            val = to_string(val)
+            (
+                yr[...],
+                day[...],
+                hr[...],
+                min[...],
+                sec[...],
+            ) = self.parse_string(val)
+
+        yrs, days, hrs, mins, secs = iterator.operands[1:]
+        jd1 = yrs * 365.25 + days + hrs / 24.0 + mins / 1440.0 + secs / 86400.0
```

Comment:
There is no reason to insist on precision loss. Since `jd1` is days anyway, just do
```
jd1 = yrs * 365.25 + days  # Exact in usual case that yrs and days are integer
jd2 = hrs / 24.0 + mins / 1440.0 + secs / 86400.0  # Inexact
self.jd1, self.jd2 = day_frac(jd1, jd2)
```

Reply:
This reduces precision loss in common cases, but `12314343.234234234234232yr` is a valid input, or even `12314343.234234234234232yr 0.1231231231212312d 1231231.12312312312s`. I don't think the latter forms are converted without loss, so we cannot guarantee no loss.

Reply:
Agreed, I just thought one might as well do the common case.

Though thinking about those longer strings, we do have the machinery for that: you can use `utils.decimal_to_twoval1` (this deals with strings too) and then `day_frac(v1, v2, factor=..., divisor=...)`. Obviously, this would only round-trip if one were to also do the reverse in output, with `twoval_to_string1`.